document-portal/file-transfer.c

```diff
@@ -0,0 +1,447 @@
+                XdpAppInfo *app_info)
+{
+  g_autoptr(GVariant) options = NULL;
+  gboolean writable;
+  gboolean autostop;
+  FileTransfer *transfer;
+  const char *sender;
+
+  g_variant_get (parameters, "(@a{sv})", &options);
+  if (!g_variant_lookup (options, "writable", "b", &writable))
+    writable = FALSE;
+
+  if (!g_variant_lookup (options, "autostop", "b", &autostop))
+    autostop = TRUE;
+
+  sender = g_dbus_method_invocation_get_sender (invocation);
+
+  transfer = file_transfer_start (app_info, sender, writable, autostop);
+
+  g_dbus_method_invocation_return_value (invocation, g_variant_new ("(s)", transfer->key));
+}
+
+static void
+add_files (GDBusMethodInvocation *invocation,
+           GVariant *parameters,
+           XdpAppInfo *app_info)
+{
+  FileTransfer *transfer;
+  const char *key;
+  g_autoptr(GVariant) options = NULL;
+  GDBusMessage *message;
+  GUnixFDList *fd_list;
+  g_autoptr(GVariantIter) iter = NULL;
+  g_autoptr(GPtrArray) files = NULL;
+  GError *error = NULL;
+  int fd_id;
+  const int *fds;
+  int n_fds;
+
+  g_variant_get (parameters, "(&sah@a{sv})", &key, &iter, &options);
+
+  transfer = lookup_transfer (key);
+  if (transfer == NULL)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid transfer");
+      return;
+    }
+
+  TRANSFER_AUTOLOCK (transfer);
+
+  if (strcmp (transfer->sender, g_dbus_method_invocation_get_sender (invocation)) != 0)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid transfer");
+      return;
+    }
+
+  message = g_dbus_method_invocation_get_message (invocation);
+  fd_list = g_dbus_message_get_unix_fd_list (message);
+
+  files = g_ptr_array_new_with_free_func (g_free);
+  fds = g_unix_fd_list_peek_fds (fd_list, &n_fds);
+
+  while (g_variant_iter_next (iter, "h", &fd_id))
+    {
+      int fd = -1;
+      g_autofree char *path = NULL;
+      gboolean fd_is_writable;
+
+      if (fd_id < n_fds)
+        fd = fds[fd_id];
+
+      if (fd == -1)
+        {
+          g_dbus_method_invocation_return_gerror (invocation, error);
+          return;
+        }
+
+      path = xdp_app_info_get_path_for_fd (app_info, fd, 0, NULL, &fd_is_writable);
+      if (path == NULL || (transfer->writable && !fd_is_writable))
+        {
+          g_dbus_method_invocation_return_error (invocation,
+                                                 XDG_DESKTOP_PORTAL_ERROR,
+                                                 XDG_DESKTOP_PORTAL_ERROR_NOT_ALLOWED,
+                                                 "Can't export file");
+          return;
+        }
+
```

Comment:
validate_fd() calls xdp_app_info_get_path_for_fd() which yes, does some validation, but then it does one extra check, on the parent directory, to make sure that the actual child name from the parent directory finds the same file.

Assume the app has rw access to /foo. Then there is an attack:
* create /foo/bar/passwd
* fd = open (/foo/bar/passwd)
* transfer.add (fd)
* portal gets the path for /foo/bar/passwd and adds it to the list of paths in transfer->files
* mv /foo/bar /foo/old-bar
* ln -s /etc /foo/bar
* transfer.execute()
* portal now opens /foo/bar/passwd, which now points to /etc/passwd, and makes a document for this
* app reads /etc/passwd

the main portal add_files() avoids this, because at the time of addition it takes the parent directory inode and verifies that the child is actually in that, then it stores the parent inode and requires all later ops to be in the same directory. However, in this new case there is a delay between these two steps, which allows this attack.

Reply:
I see. Will fix that up, then

Reply:
We now record the parent dev and ino at the time the file is added, so we can prevent this attack.